File: javascript/data-structures/p238_LinkedList.js
/*
    CLRS Section 10.2, p. 238
*/

class LinkedList {
    constructor(head = null) {
        this.head = head;
    }

    listSearch = (k) => {
        let x = this.head;
        while (x !== null && x.key !== k) {
            x = x.next;
        }
        return x;
    }

    listInsert = (x) => {
        x.next = this.head;
        if (this.head !== null) {
            this.head.prev = x;
        }
        this.head = x;
        x.prev = null;
    }

    listDelete = (x) => {
        if (x.prev !== null) {
            x.prev.next = x.next;
        } else {
            this.head = x.next;
        }
        if (x.next !== null) {
            x.next.prev = x.prev;
        }
    }
}

module.exports = {
    LinkedList
}

File: python/data_structures/p238_linked_list.py
'''
    CLRS Section 10.2, p. 238
'''

class LinkedList:
    def __init__(self, head=None):
        self.head = head
    
    def list_search(self, k):
        x = self.head
        while x is not None and x.key != k:
            x = x.next
        return x
    
    def list_insert(self, x):
        x.next = self.head
        if self.head is not None:
            self.head.prev = x
        self.head = x
        x.prev = None
    
    def list_delete(self, x):
        if x.prev is not None:
            x.prev.next = x.next
        else:
            self.head = x.next
        if x.next is not None:
            x.next.prev = x.prev
